staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go

```diff
@@ -0,0 +1,261 @@
+
+	// Resets the manager's known list of group-versions and replaces them
+	// with the given groups
+	// Thread-Safe
+	SetGroups([]apidiscoveryv2beta1.APIGroupDiscovery)
+
+	http.Handler
+}
+
+type noopResourceManager struct{}
+
+func (noopResourceManager) AddGroupVersion(groupName string, value apidiscoveryv2beta1.APIVersionDiscovery) {
+}
+func (noopResourceManager) RemoveGroup(groupName string)                      {}
+func (noopResourceManager) RemoveGroupVersion(gv metav1.GroupVersion)         {}
+func (noopResourceManager) SetGroups([]apidiscoveryv2beta1.APIGroupDiscovery) {}
+func (noopResourceManager) WebService() *restful.WebService                   { return nil }
+func (noopResourceManager) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	resp.WriteHeader(http.StatusNotImplemented)
+}
+
+var _ ResourceManager = noopResourceManager{}
+
+func NewNoopResourceManager() ResourceManager {
+	return noopResourceManager{}
+}
+
+type resourceDiscoveryManager struct {
+	serializer runtime.NegotiatedSerializer
+	// cache is an atomic pointer to avoid the use of locks
+	cache atomic.Pointer[cachedGroupList]
+
+	// Writes protected by the lock.
+	// List of all apigroups & resources indexed by the resource manager
+	lock          sync.RWMutex
+	apiGroups     map[string]*apidiscoveryv2beta1.APIGroupDiscovery
+	apiGroupNames []string
+}
+
+func NewResourceManager() ResourceManager {
+	scheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(scheme)
+	utilruntime.Must(k8sscheme.AddToScheme(scheme))
+	utilruntime.Must(apidiscoveryv2beta1.AddToScheme(scheme))
+	return &resourceDiscoveryManager{serializer: codecs}
+}
+
+func (rdm *resourceDiscoveryManager) SetGroups(groups []apidiscoveryv2beta1.APIGroupDiscovery) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	rdm.apiGroups = nil
+	rdm.apiGroupNames = nil
+	rdm.cache.Store(nil)
+
+	for _, group := range groups {
+		for _, version := range group.Versions {
+			rdm.addGroupVersionLocked(group.Name, version)
+		}
+	}
+}
+
+func (rdm *resourceDiscoveryManager) AddGroupVersion(groupName string, value apidiscoveryv2beta1.APIVersionDiscovery) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	rdm.addGroupVersionLocked(groupName, value)
+}
+
+func (rdm *resourceDiscoveryManager) addGroupVersionLocked(groupName string, value apidiscoveryv2beta1.APIVersionDiscovery) {
+	klog.Infof("Adding GroupVersion %s %s to ResourceManager", groupName, value.Version)
+
+	if rdm.apiGroups == nil {
+		rdm.apiGroups = make(map[string]*apidiscoveryv2beta1.APIGroupDiscovery)
+	}
+
+	if existing, groupExists := rdm.apiGroups[groupName]; groupExists {
+		// If this version already exists, replace it
+		versionExists := false
+
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
+				versionExists = true
+				break
+			}
+		}
+
+		if !versionExists {
```

Comment:
if the version already exists, you can return early and avoid the cache clear, right?

Reply:
If the version exists, we'd overwrite it. I guess we can add a comparison whether the old version and the new version are the same, but I'm not sure how often we'd call this function with a noop update.

Reply:
> If the version exists, we'd overwrite it. I guess we can add a comparison whether the old version and the new version are the same, but I'm not sure how often we'd call this function with a noop update.

handling no-op updates to be "free" will make every reconciliation loop very easy since they can simply call `AddGroupVersion` without having to know if it's different or not.